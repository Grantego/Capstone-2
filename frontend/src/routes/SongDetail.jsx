import React from "react";
import SpellerApi from "../../api";
import { useLoaderData, redirect } from "react-router";
import { Form } from "react-router-dom";
import { Button } from "reactstrap";
import "./songDetail.css"


export async function action({request}) {
  let formData = await request.formData()
  let song = formData.get("song")
  await SpellerApi.deleteSong(song)
  return redirect(`/${localStorage.getItem("username")}`)
}
export async function loader({params}) {
    const res = await SpellerApi.getSongDetails(params.id)
    res.song.chords = await Promise.all(res.song.chords.map(async c => {
        const {chord} = await SpellerApi.getChord(c)
        return chord
    }))
    return res
}

const SongDetail = () => {
    const {song} = useLoaderData()
    console.log(song)
    return (
        <>
         <h1>{song.title}</h1>
         <h3>By {song.username}</h3>
         <Form method="post">
            <Button type="submit" name="song" value={song.id}>Delete</Button>
         </Form>
         <div className="chord-list">
            {song.chords.map((chord, i) => (
            <div key={`${i}${chord.name}`}className="chord">
                {chord.name}
                <div className="chord-chart">
                    <img src={chord.chordChart} alt={`${chord.name} chart`} />
                </div>
            </div>                
            ))}
         </div>
        </>
    )
}


export default SongDetail