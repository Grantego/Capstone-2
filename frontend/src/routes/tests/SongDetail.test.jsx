import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import SongDetail from '../SongDetail';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect, vi } from 'vitest';
import SpellerApi from '../../../api';


vi.mock('../../../api',  () => {
    const mockChord = {name: "C", spelling: "C, E, G", chordChart: "testchart"}
    const mockSong = {
        id: 1,
        title: "Test Song",
        username: "testuser",
        chords: ['C','C','C','C']
    }
    return {
        default: vi.fn().mockImplementation(() => {
            return {
                getChord: vi.fn().mockResolvedValue({chord: mockChord}),
                getSongDetails: vi.fn().mockResolvedValue({song: mockSong})
            }
        })         
    }
});

describe('SongDetail component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();
    
    const mockLoader = async () => {
        const apiInstance = new SpellerApi();
        const res = await apiInstance.getSongDetails();
        res.song.chords = await Promise.all(res.song.chords.map(async c => {
            const {chord} = await apiInstance.getChord(c)
            return chord
        }))
        return res
    }

    router = createMemoryRouter([
        {
            path: '/',
            element: <SongDetail/>,
            loader: mockLoader
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders all elements correctly", async () => {
    localStorage.setItem('username', 'testuser')
    localStorage.setItem('token', 'sampletoken')

    render(<RouterProvider router={router} />);

    await waitFor(() => {
        // Header elements
        expect(screen.getByText("Test Song")).toBeInTheDocument()
        expect(screen.getByText("By testuser")).toBeInTheDocument()
        expect(screen.getByText("Delete")).toBeInTheDocument()
        const elements = screen.queryAllByText("C")
        expect(elements).toHaveLength(4);
    })

  })

});