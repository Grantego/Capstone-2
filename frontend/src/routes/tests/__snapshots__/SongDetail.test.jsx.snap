// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewSong component > matches snapshot 1`] = `<DocumentFragment />`;

exports[`SongDetail component > matches snapshot 1`] = `<DocumentFragment />`;
